kotlinx-coroutines-core/common/src/Delay.kt

```diff
@@ -154,8 +154,15 @@ public suspend fun delay(duration: Duration): Unit = delay(duration.toDelayMilli
 internal val CoroutineContext.delay: Delay get() = get(ContinuationInterceptor) as? Delay ?: DefaultDelay
 
 /**
- * Convert this duration to its millisecond value.
- * Positive durations are coerced at least `1`.
+ * Convert this duration to its millisecond value. Durations which have a nanosecond component less than
+ * a single millisecond will be rounded up to the next largest millisecond.
  */
-internal fun Duration.toDelayMillis(): Long =
-    if (this > Duration.ZERO) inWholeMilliseconds.coerceAtLeast(1) else 0
+internal fun Duration.toDelayMillis(): Long {
+    val millis = inWholeMilliseconds
+    val nanosecondsInMillisecond = 1_000_000
+    return when {
+        this <= Duration.ZERO -> 0L
+        millis * nanosecondsInMillisecond < inWholeNanoseconds -> millis + 1
```

Comment:
It might be more efficient to check `isFinite() && this != millis.milliseconds` since it doesn't involve any math operations (especially on JS where they're slow). Plus I suspect you're not handling overflow with the current code correctly.

Reply:
Perhaps it's best to do modulo arithmetic?